### ISSUE TITLE
Strang behavior in memory copy with control dependencies.

### ISSUE THREAD

**Issue by sleepfin** [tensorflow/tensorflow#19017]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: I have costom code.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: `redhat-7.2`
- **TensorFlow installed from (source or binary)**: `binary`
- **TensorFlow version (use command below)**: `1.4.0`
- **Python version**: `2.7`
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: `CUDA-8.0`, `cuDNN-6.0`
- **GPU model and memory**: `Nvidia - K80`
- **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
There are two different ways to apply memory copy from CPU to GPU:0 in the following code:
```
  with tf.control_dependencies([update_op]):
    # Method-1 :
    # with tf.device('/GPU:0'):
    #   a_cpu_to_gpu = a_cpu.read_value()
    # train_ops.append(vars[i].assign(a_cpu_to_gpu).op)

    # Method-2 :
    train_ops.append(vars[i].assign(a_cpu).op)
```
`a_cpu` is a variable in CPU device. `vars[i]` is a variable in GPU:0 device. The first one (Method-1) is to read the value in GPU:0 and then assign it to `a_cpu_to_gpu`. the second one (Method-2) is directly assign the variable in CPU to the variable in GPU:0.

When I enable the `trace_level`, I find that the timeline of those two codes are different:

Method-1
![image](https://user-images.githubusercontent.com/7370869/39508070-8e8204a4-4e13-11e8-97e3-f166403f0bef.png)

Method-2
![image](https://user-images.githubusercontent.com/7370869/39507750-05f705d6-4e12-11e8-8aec-ff88db046e77.png)

In `Method-1`, I think the read_value op `a_cpu.read_value()` only depends on the i-th `update_op`. So the MEMCPYHtoD (green bar) should be right after the the i-th update_op, which means right after the MEMCPYDtoH (purple bar). I don't understand why they appear in the end of the timeline.

In `Method-2`, MEMCPYHtoD (green bar) appears at the beginning of the timeline. `vars[i].assign(a_cpu).op` depends on `update_op`, how can the memcpy be executed before the `update_op` ?

I have no clue why it behaves like this, can anyone tell me why the memcpy behaves like that in `Method-1` and `Method-2` and what's the difference between them.

Thanks.

### Source code / logs
```
import tensorflow as tf
import os

from tensorflow.python.client import timeline
slim = tf.contrib.slim

train_ops = []

optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.1)

net = tf.random_normal(shape=(32, 16, 16, 128))

with tf.device('/GPU:0'):

  for i in range(10):
    net = slim.conv2d(net, 128, [11, 11], padding='SAME', scope='conv_%s' % i)

  loss = tf.reduce_mean(net, name='loss_func')
  grad = tf.gradients(loss, tf.global_variables(), gate_gradients=True, name='my_gradients')

vars = tf.global_variables()
num_vars = len(vars)

for i in range(num_vars - 1, -1, -1):

  with tf.device('/CPU:0'):
    a_cpu = tf.get_variable('a_cpu_%s' % i, initializer=vars[i].initial_value)
    update_op = optimizer.apply_gradients([(grad[i], a_cpu)], name='apply_%s' % i)

  with tf.control_dependencies([update_op]):
    # Method-1 :
    # with tf.device('/GPU:0'):
    #   a_cpu_to_gpu = a_cpu.read_value()
    # train_ops.append(vars[i].assign(a_cpu_to_gpu).op)

    # Method-2 :
    train_ops.append(vars[i].assign(a_cpu).op)


with tf.Session() as sess:
  sess.run(tf.global_variables_initializer())
  for i in range(10):
    run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
    run_metadata = tf.RunMetadata()
    sess.run(train_ops, options=run_options, run_metadata=run_metadata)
    trace_filename = os.path.join('/tmp/delete_me/trace', 'trace-global-step-%d.json' % i)
    if not tf.gfile.Exists(os.path.dirname(trace_filename)):
      os.makedirs(os.path.dirname(trace_filename))
    trace = timeline.Timeline(step_stats=run_metadata.step_stats)
    with tf.gfile.Open(trace_filename, 'w') as trace_file:
      trace_file.write(trace.generate_chrome_trace_format(show_memory=False))
```

**Comment by sleepfin** [comment]
I also tried another code which copy `a_cpu` from CPU to GPU:1 (not GPU:0)
code:
```
import tensorflow as tf
import os

from tensorflow.python.client import timeline
slim = tf.contrib.slim

train_ops = []

optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.1)

net = tf.random_normal(shape=(32, 16, 16, 128))

with tf.device('/GPU:0'):

  for i in range(10):
    net = slim.conv2d(net, 128, [11, 11], padding='SAME', scope='conv_%s' % i)

  loss = tf.reduce_mean(net, name='loss_func')
  grad = tf.gradients(loss, tf.global_variables(), gate_gradients=True, name='my_gradients')

vars = tf.trainable_variables()
num_vars = len(vars)

for i in range(num_vars - 1, -1, -1):

  with tf.device('/CPU:0'):
    a_cpu = tf.get_variable('a_cpu_%s' % i, initializer=vars[i].initial_value)
  update_op = optimizer.apply_gradients([(grad[i], a_cpu)], name='apply_%s' % i)

  with tf.control_dependencies([update_op]):
    # Method-1 :
    # with tf.device('/GPU:0'):
    #   a_cpu_to_gpu = a_cpu.read_value()
    # train_op = vars[i].assign(a_cpu_to_gpu).op
    # train_ops.append(train_op)

    # Method-2 :
    # train_ops.append(vars[i].assign(a_cpu).op)

    # Method-3 :
    with tf.device('/GPU:1'):
      a_cpu_to_gpu = a_cpu.read_value()
    train_ops.append(a_cpu_to_gpu.op)

with tf.Session() as sess:
  sess.run(tf.global_variables_initializer())
  for i in range(10):
    run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
    run_metadata = tf.RunMetadata()
    sess.run(train_ops, options=run_options, run_metadata=run_metadata)
    trace_filename = os.path.join('/tmp/delete_me/trace', 'trace-global-step-%d.json' % i)
    if not tf.gfile.Exists(os.path.dirname(trace_filename)):
      os.makedirs(os.path.dirname(trace_filename))
    trace = timeline.Timeline(step_stats=run_metadata.step_stats)
    with tf.gfile.Open(trace_filename, 'w') as trace_file:
      trace_file.write(trace.generate_chrome_trace_format(show_memory=False))
```

![image](https://user-images.githubusercontent.com/7370869/39515871-ee9d8664-4e2d-11e8-9484-d3e6d6836760.png)


In this way, MEMCPYHtoD (green bar) overlaps with compute_gradients.
But in `Method-2`, MEMCPYHtoD (green bar) can not overlap with compute_gradients.

It seems that MEMCPYHtoD (green bar) cannot overlap with the backward computation. But it can somehow overlap with forward computation (blue bars) in `Method-1`.

If I change the `with tf.device('/GPU:1'):` to `with tf.device('/GPU:0'):`, the timeline is:

![image](https://user-images.githubusercontent.com/7370869/39515848-d8956864-4e2d-11e8-9d5d-f39fc6148a74.png)

**Comment by yaroslavvb** [comment]
Control dependencies only affect operations created within the control dependencies block. Because `a_cpu` was created outside of that block, it can evaluate before `update_op` has executed. `a_cpu_to_gpu` is created within the block, so it's guaranteed to run after update_op

**Comment by sleepfin** [comment]
I have simplified the code to :
```
import tensorflow as tf
import os

from tensorflow.python.client import timeline

train_ops = []
endpoints = []

with tf.device('/GPU:0'):
  a = tf.random_normal(shape=(2048, 2048), name='a')
  b = tf.random_normal(shape=(2048, 2048), name='b')
  for i in range(10):
    a = tf.matmul(a, b, name='matmul_a_b_%s' % i)
    endpoints.append(a)

for i in range(len(endpoints)):
  with tf.device('/CPU:0'):
    memcpy_d2h = tf.identity(endpoints[i])
  with tf.device('/GPU:0'):
    memcpy_h2d = tf.identity(memcpy_d2h)
  train_ops.append(memcpy_h2d.op)


with tf.Session() as sess:
  sess.run(tf.global_variables_initializer())
  for i in range(10):
    run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
    run_metadata = tf.RunMetadata()
    sess.run(train_ops, options=run_options, run_metadata=run_metadata)
    trace_filename = os.path.join('/tmp/delete_me/trace', 'trace-%d.json' % i)
    if not tf.gfile.Exists(os.path.dirname(trace_filename)):
      os.makedirs(os.path.dirname(trace_filename))
    trace = timeline.Timeline(step_stats=run_metadata.step_stats)
    with tf.gfile.Open(trace_filename, 'w') as trace_file:
      trace_file.write(trace.generate_chrome_trace_format(show_memory=False))
```

This code outputs the same timeline as Method-1 (all MEMCPYHtoD ops execute after the last MEMCPYDtoH)
and also when I change the device of from GPU:0 to GPU:1
```
  with tf.device('/GPU:1'):
    memcpy_h2d = tf.identity(memcpy_d2h)
```
It outputs the same timeline as Method-3 (MEMCPYDtoH and MEMCPYHtoD execute alternately and sequentially.)

In addition, I also reverse the device GPU and CPU like:
```
with tf.device('/GPU:0'):
  a = tf.random_normal(shape=(2048, 2048), name='a')
  b = tf.random_normal(shape=(2048, 2048), name='b')
  for i in range(10):
    a = tf.matmul(a, b, name='matmul_a_b_%s' % i)
    endpoints.append(a)

for i in range(len(endpoints)):
  with tf.device('/CPU:0'):
    memcpy_d2h = tf.identity(endpoints[i])
  with tf.device('/GPU:0'):
    memcpy_h2d = tf.identity(memcpy_d2h)
  train_ops.append(memcpy_h2d.op)
```
I have timeline:
![image](https://user-images.githubusercontent.com/7370869/39563617-8fe18ff6-4ee3-11e8-9380-8c7f9049cdfc.png)


In this case, MEMCPYDtoH and MEMCPYHtoD execute alternately and sequentially.
What is reason of these differences?

**Comment by sleepfin** [comment]
Well, I made some more experiments. If I change the size of `a` and `b`, I find that `Matmul` can somehow overlap with `MEMCPYH2D`
When `a` and `b` are 512 x 512
```
import tensorflow as tf
import os

from tensorflow.python.client import timeline

train_ops = []
endpoints = []

with tf.device('/GPU:0'):
  a = tf.random_normal(shape=(512, 512), name='a')
  b = tf.random_normal(shape=(512, 512), name='b')
  for i in range(10):
    a = tf.matmul(a, b, name='matmul_a_b_%s' % i)
    endpoints.append(a)

for i in range(len(endpoints)):
  with tf.device('/CPU:0'):
    memcpy_d2h = tf.identity(endpoints[i])
  with tf.device('/GPU:0'):
    memcpy_h2d = tf.identity(memcpy_d2h)
  train_ops.append(memcpy_h2d.op)


with tf.Session() as sess:
  sess.run(tf.global_variables_initializer())
  for i in range(10):
    run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
    run_metadata = tf.RunMetadata()
    sess.run(train_ops, options=run_options, run_metadata=run_metadata)
    trace_filename = os.path.join('/tmp/delete_me/trace', 'trace-%d.json' % i)
    if not tf.gfile.Exists(os.path.dirname(trace_filename)):
      os.makedirs(os.path.dirname(trace_filename))
    trace = timeline.Timeline(step_stats=run_metadata.step_stats)
    with tf.gfile.Open(trace_filename, 'w') as trace_file:
      trace_file.write(trace.generate_chrome_trace_format(show_memory=False))
```

Timeline is:
![image](https://user-images.githubusercontent.com/7370869/39565097-54d615ca-4ee9-11e8-9066-b45e4cc7ffcf.png)

When When `a` and `b` are 256 x 256:
```
  a = tf.random_normal(shape=(256, 256), name='a')
  b = tf.random_normal(shape=(256, 256), name='b')
```
Timeline is:
![image](https://user-images.githubusercontent.com/7370869/39565157-8817f1a6-4ee9-11e8-9cf2-021d5edcdf65.png)

**Comment by sleepfin** [comment]
Anyone can help me :) @angersson @yaroslavvb

**Comment by yaroslavvb** [comment]
It's not clear what you are saying is a bug, can you isolate the problem further?

**Comment by sleepfin** [comment]
@yaroslavvb My question is in Method-1, why MEMCPYH2D can not overlap with the backward computation while MEMCPYD2H can.

**Comment by angersson** [comment]
(I marked @yaroslavvb as assignee since he's been looking at this already. Thanks!)

**Comment by sleepfin** [comment]
Still waiting for reply :-)

**Comment by sleepfin** [comment]
Still waiting for reply :-)

**Comment by tensorflowbutler** [comment]
Nagging Assignees @yaroslavvb, @angersson: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @yaroslavvb, @angersson: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @yaroslavvb, @angersson: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @yaroslavvb, @angersson: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @yaroslavvb, @angersson: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sleepfin** [comment]
@yaroslavvb @angersson anyone help? :)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yaroslavvb: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yaroslavvb: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by alextp** [comment]
Sorry, I don't have context and I see a lot of discussion in this bug. Can you restate what is the actual issue that is still unresolved?

Also, I see variables and control dependencies; does enabling resource variables (tf.enable_resource_variables() on nightly) make the issue go away?

**Comment by alextp** [comment]
@sleepfin can you update the bug on what is the current issue? Did you check whether resource variables make this issue go away?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by wt-huang** [comment]
Closing due to low activity
